src/nvim/api/private/helpers.c

```diff
@@ -1342,3 +1346,199 @@ const char *get_default_stl_hl(win_T *wp)
+    addr_type_arg = ADDR_LINES;
+  } else if (HAS_KEY(opts->range)) {
+    api_set_error(err, kErrorTypeValidation, "Invalid value for 'range'");
+    return;
+  }
+
+  if (opts->count.type == kObjectTypeBoolean) {
+    if (opts->count.data.boolean) {
+      argt |= EX_COUNT | EX_ZEROR | EX_RANGE;
+      addr_type_arg = ADDR_OTHER;
+      def = 0;
+    }
+  } else if (opts->count.type == kObjectTypeInteger) {
+    argt |= EX_COUNT | EX_ZEROR | EX_RANGE;
+    addr_type_arg = ADDR_OTHER;
+    def = opts->count.data.integer;
+  } else if (HAS_KEY(opts->count)) {
+    api_set_error(err, kErrorTypeValidation, "Invalid value for 'count'");
+    return;
+  }
+
+  if (opts->addr.type == kObjectTypeString) {
+    if (parse_addr_type_arg((char_u *)opts->addr.data.string.data, (int)opts->addr.data.string.size,
+                            &addr_type_arg) != OK) {
+      api_set_error(err, kErrorTypeValidation, "Invalid value for 'addr'");
+      return;
+    }
+
+    if (addr_type_arg != ADDR_LINES) {
+      argt |= EX_ZEROR;
+    }
+  } else if (HAS_KEY(opts->addr)) {
+    api_set_error(err, kErrorTypeValidation, "Invalid value for 'addr'");
+    return;
+  }
+
+  if (api_object_to_bool(opts->bang, "bang", false, err)) {
+    if (ERROR_SET(err)) {
+      return;
+    }
+
+    argt |= EX_BANG;
+  }
+
+  if (api_object_to_bool(opts->bar, "bar", false, err)) {
+    if (ERROR_SET(err)) {
+      return;
+    }
+
+    argt |= EX_TRLBAR;
+  }
+
+  if (api_object_to_bool(opts->register_, "register", false, err)) {
+    if (ERROR_SET(err)) {
+      return;
+    }
+
+    argt |= EX_REGSTR;
+  }
+
+  bool force = api_object_to_bool(opts->force, "force", false, err);
+  if (ERROR_SET(err)) {
+    return;
+  }
+
+  if (opts->complete.type == kObjectTypeLuaRef) {
+    compl = EXPAND_USER_LUA;
+    compl_luaref = api_new_luaref(opts->complete.data.luaref);
+  } else if (opts->complete.type == kObjectTypeString) {
+    if (parse_compl_arg((char_u *)opts->complete.data.string.data,
+                        (int)opts->complete.data.string.size, &compl, &argt,
+                        (char_u **)&compl_arg) != OK) {
+      api_set_error(err, kErrorTypeValidation, "Invalid value for 'complete'");
+      goto out;
+    }
+  } else if (HAS_KEY(opts->complete)) {
+    api_set_error(err, kErrorTypeValidation, "Invalid value for 'complete'");
+    goto out;
+  }
+
+  switch (command.type) {
+  case kObjectTypeLuaRef:
+    luaref = api_new_luaref(command.data.luaref);
+    if (opts->desc.type == kObjectTypeString) {
+      rep = opts->desc.data.string.data;
+    } else {
+      snprintf((char *)IObuff, IOSIZE, "<Lua function %d>", luaref);
+      rep = (char *)IObuff;
+    }
+    break;
+  case kObjectTypeString:
+    rep = command.data.string.data;
+    break;
+  default:
+    api_set_error(err, kErrorTypeValidation, "'command' must be a string or Lua function");
+    goto out;
+  }
+
+  buf_T *target_buf = find_buffer_by_handle(buffer, err);
+  buf_T *save_curbuf = NULL;
```

Comment:
Let's just do `buf_T *save_curbuf = curbuf; curbuf = target_buf`. If there's no side-effects there is no need for an if.

Reply:
I think I am actually going to move all of the buffer stuff into `nvim_buf_add_user_command` directly, which should simplify this whole thing.